File: hal_interfaces/hal_interfaces/specific/amazon_warehouse/simtime.py
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy
import rosgraph_msgs.msg

### AUXILIARY FUNCTIONS
class SimTime():
    def __init__(self):
        self.sec = 0
        self.nanosec = 0
    
    def __str__(self):
        s = (
            "SimTime: {\n   sec:" 
            + str(self.sec) 
            + "\n   nanosec: " 
            + str(self.nanosec) 
            + "\n}"
            )
        
        return s

def clock2SimTime(clock):
    """
    Translates from ROS Clock to JderobotTypes SimTime.
    @param clock: ROS Clock to translate
    @type clock: Clock
    @return a SimTime translated from clock
    """
    simtime = SimTime()
    simtime.sec = clock.clock.sec
    simtime.nanosec = clock.clock.nanosec

    return simtime
    
### HAL INTERFACE ###
class SimTimeNode(Node):
    def __init__(self, topic):
        super().__init__("simtime_node")

        qos_profile = QoSProfile(
            reliability=ReliabilityPolicy.BEST_EFFORT,
            history=HistoryPolicy.KEEP_LAST,
            depth=1
        )

        self.sub = self.create_subscription(
            rosgraph_msgs.msg.Clock, 
            topic, 
            self.listener_callback, 
            qos_profile=qos_profile
        )

        self.last_simtime_ = rosgraph_msgs.msg.Clock()

    def listener_callback(self, clock):
        self.last_simtime_ = clock
    
    def getSimTime(self):
        return clock2SimTime(self.last_simtime_)

File: common_interfaces_cpp/src/hal/sim_time.cpp
#include "common_interfaces_cpp/hal/sim_time.hpp"
#include <sstream>

SimTimeData::SimTimeData() {
    seconds = 0;
    nanoseconds = 0;
}

std::string SimTimeData::to_string() const {
    std::ostringstream oss;
    oss << "SimTimeData: {\n   sec: " << seconds
        << "\n   nanosec: " << nanoseconds << "\n}";
    return oss.str();
}

double SimTimeData::to_double() const {
    return static_cast<double>(seconds) + (static_cast<double>(nanoseconds) * 1e-9);
}

SimTimeData simTime2SimTimeData(const rosgraph_msgs::msg::Clock& clock) {
    SimTimeData clockData;
    clockData.seconds = clock.clock.sec;
    clockData.nanoseconds = clock.clock.nanosec;
    return clockData;
}

SimTimeNode::SimTimeNode(const std::string& node_name) 
    : Node(node_name) {
    
    rclcpp::QoS qos_policy(rclcpp::KeepLast(1));
    qos_policy.best_effort();

    sub = this->create_subscription<rosgraph_msgs::msg::Clock>(
        "/clock", 
        qos_policy, 
        std::bind(&SimTimeNode::listener_callback, this, std::placeholders::_1)
    );
}

void SimTimeNode::listener_callback(const rosgraph_msgs::msg::Clock::SharedPtr msg) {
    std::lock_guard<std::mutex> guard(lock_);
    last_sim_time_ = *msg;
}

SimTimeData SimTimeNode::getSimTime() {
    std::lock_guard<std::mutex> guard(lock_);
    return simTime2SimTimeData(last_sim_time_);
}